File: algo_2021/s6.-dfs-comb-perm/816.Traveling_Salesman_Problem/816_traveling_salesman_problem__bf.py
class Result:
    def __init__(self):
        self.min_cost = float('inf')
class Solution:
    """
    @param n: an integer,denote the number of cities
    @param roads: a list of three-tuples,denote the road between cities
    @return: return the minimum cost to travel all cities
    """
    def minCost(self, n, roads):
        # Write your code here
        graph = self.construct_graph(roads, n)
        res = Result()
        self.dfs(1, n, set([1]), 0, graph, res)
        return res.min_cost
    
    def dfs(self, city, n, visited, cost, graph, res):
        # once traversed all the cities, update the result
        if len(visited) == n:
            res.min_cost = min(res.min_cost, cost)  
            return
        
        # traverse all the city's route
        for next_city in graph[city]:
            if next_city in visited:
                continue
            visited.add(next_city)
            self.dfs(next_city, n, visited, cost + graph[city][next_city], graph, res)
            visited.remove(next_city)          
    
    def construct_graph(self, roads, n):
        graph = {
            i : {j : float('inf') for j in range(1, n + 1)}
            for i in range(1, n + 1)
        }
        for city_1, city_2, cost in roads:
            # since there might be several different cost btw two cities
            # should pick the minimum one
            graph[city_1][city_2] = min(graph[city_1][city_2], cost)
            graph[city_2][city_1] = min(graph[city_2][city_1], cost)
        return graph

File: lintcode_algo_2021/s6.-dfs-comb-perm/816.Traveling_Salesman_Problem/816_traveling_salesman_problem__bf.java
class Result {
    int minCost;
    public Result() {
        minCost = Integer.MAX_VALUE;
    }
}

public class Solution {
    /**
     * @param n: an integer,denote the number of cities
     * @param roads: a list of three-tuples,denote the road between cities
     * @return: return the minimum cost to travel all cities
     */
    public int minCost(int n, int[][] roads) {
        // Write your code here
        int[][]graph = constructGraph(roads, n);
        Set<Integer> visited = new HashSet<>();
        Result res = new Result();
        visited.add(1);
        dfs(1, n, visited, 0, graph, res);
        return res.minCost;
    }

    private void dfs(int city,
                     int n, 
                     Set<Integer> visited,
                     int cost,
                     int[][] graph,
                     Result res) {
        if (visited.size() == n) {
            res.minCost = Math.min(res.minCost, cost);
            return;
        }
        
     // Here would traverse all the route
        for ( int i = 1; i < graph[city].length; i++) {
            if (visited.contains(i)) {
                continue;
            }
            visited.add(i);
            dfs(i, n, visited, cost + graph[city][i], graph, res);
            visited.remove(i);
        }
    }

    int[][] constructGraph(int[][] roads, int n) {
        int[][] graph = new int[n + 1][n + 1];
        for (int i = 0; i < n + 1; i++) {
            for (int j = 0; j < n + 1; j++) {
                // since later would calculate the total sum
                // those default value (MAX_VALUE) would be added as well
                // to avoid overflow, right shift 4 bits
                graph[i][j] = Integer.MAX_VALUE >> 4;
            }
        }
        for (int i = 0; i < roads.length; i++) {
            int city1 = roads[i][0];
            int city2 = roads[i][1];
            int cost = roads[i][2];
            graph[city1][city2] = Math.min(graph[city1][city2], cost);
            graph[city2][city1] = Math.min(graph[city2][city1], cost);
        }

        return graph;
    }
}